Avia HX711 driver and example

GeorgeGkinis opened the pull request:
This is a continuation of PR #[24295](https://github.com/zephyrproject-rtos/zephyr/pull/24295).

All review comments have been addressed with  the exception of [this one](https://github.com/zephyrproject-rtos/zephyr/pull/24295#pullrequestreview-590524238).

Additionally the following have been added:
- Support for **multiple hx711 instances**.
- Support for fractional **HX711_CALIBRATION_WEIGHT**
- Support for fractional **HX711_SLOPE**
- **hx711_channel_get** now returns a **sensor_value** with **val2** populated (fractional part)

Special thanks for the contribution of [@nobodyguy](https://github.com/nobodyguy/zephyr/tree/v2.7-HX711):
- Proper power management has been implemented.
- Usage of semaphores and interrupts while waiting for an available measurement.

While the merge window has now been closed (rc1) the release process [allows for the addition of previously unsupported hardware](https://docs.zephyrproject.org/latest/development_process/release_process.html):
_"An occasional exception is made for drivers for previously unsupported hardware; if they do not touch any other in-tree code, they cannot cause regressions and should be safe to add at any time"_

Thanks for the support so far and sorry for the delay, I am just a hobbyist.
Peace, Love and Freedom.

GeorgeGkinis commented:
@MaureenHelm Could you please approve the remaining workflows?

bbilas commented on drivers/sensor/hx711/hx711.h:
```diff
@@ -0,0 +1,58 @@
+	int instance;
+	bool rate_enabled;
+
+	gpio_pin_t dout_pin;
```
You should use https://docs.zephyrproject.org/latest/reference/peripherals/gpio.html#c.gpio_dt_spec and appropriate `GPIO_DT_SPEC_INST_GET` macro instead.

GeorgeGkinis replied:
done.

bbilas replied:
Now you should use `gpio_pin_X_dt` functions instead and get rid of 
```
const struct device *dout_gpio;
const struct device *sck_gpio;
const struct device *rate_gpio;
```
that are not needed anymore. Please see https://docs.zephyrproject.org/latest/reference/peripherals/gpio.html

GeorgeGkinis replied:
Done.

bbilas commented on samples/sensor/hx711/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_HX711_CALIBRATION_WEIGHT_INTEGER=500
+CONFIG_HX711_CALIBRATION_WEIGHT_DECIMAL=500000
+CONFIG_LOG=y
+CONFIG_REBOOT=y
```
Missing nl.

GeorgeGkinis replied:
done.

bbilas commented on drivers/sensor/hx711/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_HX711 hx711.c)
```
Missing nl.

tejlmand commented on drivers/sensor/hx711/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	default HX711_RATE_10HZ
+	help
+	  Select sampling rate.
+config HX711_RATE_10HZ
```
please add empty line for readability.
```suggestion

config HX711_RATE_10HZ
```

GeorgeGkinis replied:
Done.

tejlmand commented on drivers/sensor/hx711/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	  Select sampling rate.
+config HX711_RATE_10HZ
+	bool "10Hz"
+config HX711_RATE_80HZ
```
please add empty line for readability.
```suggestion

config HX711_RATE_80HZ
```

GeorgeGkinis replied:
Done.

tejlmand commented on drivers/sensor/hx711/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	default HX711_GAIN_128X
+	help
+	  Select HX711 gain.
+config HX711_GAIN_32X
```
same wrt empty line between choice entries here.

GeorgeGkinis replied:
Done.

tejlmand commented on samples/sensor/hx711/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2020, George Gkinis
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
```suggestion
cmake_minimum_required(VERSION 3.20.0)
```

GeorgeGkinis replied:
Done.

tejlmand reviewed (commented):
some minor nits.

gmarull commented on drivers/sensor/hx711/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	default 1
+	default 1 if HX711_GAIN_128X
+	default 2 if HX711_GAIN_32X
+	default 3 if HX711_GAIN_64X
```
Please move Kconfig parameters to DT

GeorgeGkinis replied:
This is not a HW configuration and can change in runtime.
Why should it be in DT?

gmarull replied:
FYI https://docs.zephyrproject.org/latest/guides/dts/dt-vs-kconfig.html#devicetree-versus-kconfig

GeorgeGkinis replied:
As stated in the documentation initial configuration parameters should reside in DT.
Fixed and added getters and settters for all parameters.

GeorgeGkinis replied:
Done.

gmarull commented on drivers/sensor/hx711/hx711.h:
```diff
@@ -0,0 +1,50 @@
+#include <drivers/sensor/hx711.h>
+
+struct hx711_data {
+	const struct device *dout_gpio;
+	const struct device *sck_gpio;
+	const struct device *rate_gpio;
```
this is not required when using `gpio_dt_spec`

GeorgeGkinis replied:
Done.

gmarull commented on include/drivers/sensor/hx711.h:
```diff
@@ -0,0 +1,79 @@
+					 const struct sensor_value *target,
+					 uint8_t readings);
+
+/**
+ * @brief Set the HX711 power.
+ *
+ * @param dev Pointer to the hx711 device structure
+ * @param power one of HX711_POWER_OFF, HX711_POWER_ON
+ * @retval The current power state or ENOTSUP if an invalid value pow is given
+ *
+ */
+int avia_hx711_power(const struct device *dev, enum hx711_power power);
```
please add support for device PM, do not implement a custom API.

GeorgeGkinis replied:
Fixed.

GeorgeGkinis commented:
I see that [checkpatch ](https://github.com/zephyrproject-rtos/zephyr/runs/5109723191?check_suite_focus=true) fails here :

```
-:236: WARNING:LONG_LINE_STRING: line length of 104 exceeds 100 columns
#236: FILE: drivers/sensor/hx711/hx711.c:95:
+			LOG_ERR("Weight data not ready within %d ms. Is the device properly connected?",

- total: 0 errors, 1 warnings, 1184 lines checked
```

but if I select the line:
'			LOG_ERR("Weight data not ready within %d ms. Is the device properly connected?",'
 my editor counts 83 characters.

Are tabs replaced by spaces during checkpatch?

gmarull commented on samples/sensor/hx711/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_HX711_CALIBRATION_WEIGHT_INTEGER=500
+CONFIG_HX711_CALIBRATION_WEIGHT_DECIMAL=500000
+CONFIG_LOG=y
+CONFIG_REBOOT=y
```
if you need PM, you'll need to turn on `PM_DEVICE` and `PM_DEVICE_RUNTIME` in this case.

GeorgeGkinis replied:
Done.

gmarull commented on samples/sensor/hx711/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+  sample.sensor.hx711:
+    tags: sensors
+    depends_on: gpio
+    platform_allow: nrf52dk_nrf52832
```
please, filter by dt compat enabled: `filter: dt_compat_enabled("...")`, this sample doesn't seem to be board specific.

GeorgeGkinis replied:
Done.

gmarull commented on samples/sensor/hx711/src/main.c:
```diff
@@ -0,0 +1,142 @@
+
+const struct device *hx711;
+
+int reset(void)
+{
+	int ret = 0;
+
+	LOG_INF("Setting HX711_POWER_OFF");
+	ret = pm_device_runtime_put(hx711);
+	if (ret < 0) {
+		return ret;
+	}
+	k_msleep(1000);
+	LOG_INF("Setting HX711_POWER_ON");
+	ret = pm_device_runtime_get(hx711);
+	if (ret < 0) {
+		return ret;
+	}
+	return ret;
+}
```
this should be done by the driver itself, see my previous comments.

GeorgeGkinis replied:
Removed and works.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+ * @return void
+ *
+ */
+static void hx711_attr_set_slope(struct hx711_data *data, const struct sensor_value *slope)
+{
+	data->slope.val1 = slope->val1;
+	data->slope.val2 = slope->val2;
+}
```
done.

I used to not support fractional parts.
I just added a line instead of simplifying.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+	irq_unlock(key);
+#endif
+
+	count ^= 0x800000;
```
Defined in the datasheet C sample page 8 : https://datasheetspdf.com/pdf-file/842201/Aviasemiconductor/HX711/1
Linus does the same : https://github.com/torvalds/linux/blob/e6251ab4551f51fa4cee03523e08051898c3ce82/drivers/iio/adc/hx711.c#L157

^= 0x800000 adds padding to make a 32 bit int from a 24bit one.

Comment added.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+ * @return void
+ *
+ */
+static void hx711_attr_set_offset(struct hx711_data *data, const struct sensor_value *offset)
+{
+	data->offset = offset->val1;
+}
+
+/**
+ * @brief Set HX711 slope.
+ *
+ * @param data Pointer to the hx711 driver data structure
+ * @param slope sensor_value struct. Slope used to calculate the real weight.
+ *        Formula is :
+ *          weight = slope * (reading - offset)
+ *
+ * @return void
+ *
+ */
+static void hx711_attr_set_slope(struct hx711_data *data, const struct sensor_value *slope)
+{
+	data->slope.val1 = slope->val1;
+	data->slope.val2 = slope->val2;
+}
```
Nice :)
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+
+	switch (chan) {
+	case SENSOR_CHAN_WEIGHT:
+	case SENSOR_CHAN_ALL: {
```
I looked it up in the documentation I found it mentioned that **sensor_channel_get** should 
"..Return a useful value for a **particular** channel .."
[sensor_channel_get](https://docs.zephyrproject.org/latest/reference/peripherals/sensor.html?highlight=channel_get#c.sensor_channel_get).

Since the function signature is : 
_int sensor_channel_get(const struct device *dev, **enum sensor_channel chan**, struct sensor_value *val)_
and **SENSOR_CHAN_ALL** is of type **enum sensor_channel** then **SENSOR_CHAN_ALL** is accepted as a value.

The current function implementation accepts both and rejects any other invalid channel.

**SENSOR_CHAN_WEIGHT** is also subset of **SENSOR_CHAN_ALL** and in this particular case also equal.
It would seem strange from the user perspective to ask for all channels on a sensor which has only one and get an error instead of a valid **sensor_value**.

Of course it would be an issue for sensors with multiple channels. The API should then return a pointer to an sensor_value array or something which is not currently the case.

I understand your argument though.
It _should_ only be supported when fetching but is currently not ;)

GeorgeGkinis replied:
Removed case SENSOR_CHAN_ALL
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+		.instance = index,				      \
+		.dout = GPIO_DT_SPEC_INST_GET(index, dout_gpios),     \
+		.sck = GPIO_DT_SPEC_INST_GET(index, sck_gpios),	      \
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(index, rate_gpios), \
+			    (HX711_RATE_CFG(index)),		      \
+			    (.rate_enabled = false))		      \
```
I like this one!
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+	struct hx711_data *data = dev->data;
+	const struct hx711_config *cfg = dev->config;
+
+	if (data->power != HX711_POWER_ON) {
+		return -EACCES;
+	}
```
Fixed in hx711_sample_fetch. 

HX711 needs a quite long time though to provide the first measurement after wake-up (up to 600~700ms).
How could a user keep the sensor "on" if he needs quick measurements?

I need to think a bit more about this and read more.

GeorgeGkinis replied:
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+#ifdef CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING
+	uint32_t key = irq_lock();
+#endif
+	for (i = 0; i < 24; i++) {
```
Added comment : 
`	/* HX711 is an 24bit ADC.
	 * Clock out 24 bits.*/`

GeorgeGkinis replied:
Done.

GeorgeGkinis commented on samples/sensor/hx711/src/main.c:
```diff
@@ -0,0 +1,142 @@
+{
+	int ret = 0;
+
+	hx711 = device_get_binding("HX711");
+
+	__ASSERT(hx711, "Failed to get device binding");
+
+	LOG_INF("device is %p, name is %s", hx711, hx711->name);
```
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.h:
```diff
@@ -0,0 +1,47 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_SENSOR_HX711_H_
```
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,584 @@
+
+	return ret;
+}
+#endif /* CONFIG_PM_DEVICE */
```
Good one.
Done.

GeorgeGkinis commented:
@gmarull Thanks for the time you took to provide feedback.
All your suggestions where appreciated and fixed.

I suppose in the future I should push a commit per review item, so that it does not become "outdated" and to ease verification of the changes ?
After all are resolved then I should squash them an acceptable commit(s)?

mbolivar-nordic commented on drivers/sensor/hx711/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_HX711 hx711.c)
```
Please just use `zephyr_library_sources(hx711.c)`. This style is redundant and is undoing the cleanup work that was done in bd8afe9365e19d579417a514dd3bc87ee1667b9f

GeorgeGkinis replied:
Done.

mbolivar-nordic commented on dts/bindings/sensor/avia,hx711.yaml:
```diff
@@ -0,0 +1,90 @@
+  offset:
+    type: int
+    required: false
+    default: 0
```
All use of defaults in bindings must be justified in the descriptions. Please read https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings and update this PR accordingly.

GeorgeGkinis replied:
Thanks for your review @mbolivar-nordic.

I cannot think of a justification for **offset, slope** , and **calibration-weight**.

I just want those values to have an initial known value instead of garbage in case the user wants to set them up at runtime and not define them in his overlay file.

Would it be acceptable to use **DT_INST_PROP_OR** instead? : 
```
#define HX711_INIT(index) \
    const int32_t  def[2]={1,0};                 \
	static struct hx711_data hx711_data_##index = {				      \
		.reading = 0,							      \
		.offset = DT_INST_PROP_OR(index, offset, 0),				      \
		.cw = DT_INST_PROP_OR(index, calibration_weight, &def),			      \
		.sl = DT_INST_PROP_OR(index, slope, &def),				      \
		.sample_fetch_timeout = DT_INST_PROP(index, sample_fetch_timeout_ms), \
		.gain = DT_INST_PROP(index, gain),				      \
		.rate = DT_INST_PROP_OR(index, rate_hz,HX711_RATE_10HZ),				      \
	};									      \
```

For **sample-fetch-timeout-ms** and **gain** there is justification based on the sensor itself: time to first reading and gain after power on.

GeorgeGkinis replied:
I found a way to justify the default values for offset, slope and calibration weight.

GeorgeGkinis replied:
@mbolivar-nordic please check if the latest changes are acceptable

mbolivar-nordic reviewed (changes requested):
Thanks for the driver. A couple of requests from the DT and build system side.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,623 @@
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <sys/byteorder.h>
```
Done. 
Also for example main.c

MaureenHelm commented:
> Is there anything I should do at this point? I have requested everyone that provided feedback to review again.
> 

No, just wait for re-reviews

> I also see that v3.0.0 stable is out. I worked from master, so fetch --all and rebase should do the trick?

Please stay on the latest main branch, don't rebase to the release branch.

MaureenHelm commented on samples/sensor/hx711/src/main.c:
```diff
@@ -0,0 +1,93 @@
+
+void set_rate(enum hx711_rate rate)
+{
+	static struct sensor_value rateVal;
```
Please don't use camel case

GeorgeGkinis replied:
Done.

MaureenHelm commented:
Please rebase to the latest main to pick up fixes for the failed footprint and documentation checks. The failed compliance check identified some real issues that you need to fix:
```Error: See docs.zephyrproject.org/latest/contribute/#coding-style for more details.

Missing newline at end of 'dts/bindings/sensor/avia,hx711.yaml'. Check your text editor settings.
Error: See docs.zephyrproject.org/latest/guides/kconfig/index.html for more details.

Found pointless 'menuconfig' symbols without children. Use regular 'config'
symbols instead. See
docs.zephyrproject.org/latest/guides/kconfig/tips.html#menuconfig-symbols.

HX711                               drivers/sensor/hx711/Kconfig:5


Found references to undefined Kconfig symbols. If any of these are false
positives, then add them to UNDEF_KCONFIG_WHITELIST in check_compliance.py in the ci-tools repo.

If the reference is for a comment like /* CONFIG_FOO_* */ (or
/* CONFIG_FOO_*_... */), then please use exactly that form (with the '*'). The
CI check knows not to flag it.

More generally, a reference followed by $, @, {, *, or ## will never be
flagged.

CONFIG_HX711_DISABLE_INTERRUPTS_WHILE_POLLING drivers/sensor/hx711/hx711.c:129, drivers/sensor/hx711/hx711.c:147

```

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,622 @@
+	int sample_fetch_timeout;
+	char gain;
+	enum hx711_rate rate;
+	const int32_t cw[2];
+	const int32_t sl[2];
```
Done.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,622 @@
+
+	/* SCK set LOW */
+	gpio_pin_set_dt(&cfg->sck, false);
+	k_busy_wait(1);
```
From the datasheet page 5 : PD_SCK high time = PD_SCK low time = 1μs (TYP)

MaureenHelm replied:
Please add a comment to document this in the code

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,622 @@
+		}
+	}
+
+	/* Clock data out. Optionally disable interrupts */
```
For the rare case that some application finds measuring weight more important than serving other routines.
On second thought this might just better be removed.

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,622 @@
+	default:
+		return -ENOTSUP;
+	}
+	return hx711_sample_fetch(dev, SENSOR_CHAN_WEIGHT);
```
Setting gain involves taking a reading and adding additional pulses at the end of the reading.
See Datasheet page 4 : 
_"Input and gain selection is controlled by the
number of the input PD_SCK pulses (Table 3).
PD_SCK clock pulses should not be less than 25
or more than 27 within one conversion period, to
avoid causing serial communication error. "_

hx711_sample_fetch() reads data->gain and does that on line 143

MaureenHelm replied:
Please add the explanation to the code comment

GeorgeGkinis commented:
@MaureenHelm I seem to have trouble runing scripts/kconfig/lint.py : 
```
q@q-hole:~/zephyr-projects/hx711-driver/zephyr$ source zephyr-env.sh 
q@q-hole:~/zephyr-projects/hx711-driver/zephyr$ scripts/kconfig/lint.py 
scripts/kconfig/lint.py: modules/Kconfig:6: '/Kconfig.modules' not found (in 'source "$(KCONFIG_BINARY_DIR)/Kconfig.modules"'). Check that environment variables are set correctly (e.g. $srctree, which is set to 'scripts/kconfig/../..'). Also note that unset environment variables expand to the empty string.
```

I believe the warnings should be gone now, but would like to be able to check myself.

MaureenHelm reviewed (commented):
@mbolivar-nordic @gmarull please revisit

mbolivar-nordic reviewed (changes requested):
One of the justifications for `default` is excellent, and I will comment below exactly why. However, the rest still don't make sense to me, sorry.

Here are the relevant requirements from the documentation I linked:

```
There is a risk in using default: when the value in the binding may be incorrect for a particular board or hardware configuration. For example, defaulting the capacity of the connected power cell in a charging IC binding is likely to be incorrect. For such properties it’s better to make the property required: true, forcing the devicetree maintainer into an explicit and witting choice.

Driver developers should use their best judgment as to whether a value can be safely defaulted. Candidates for default values include:

    delays that would be different only under unusual conditions (such as intervening hardware)

    configuration for devices that have a standard initial configuration (such as a USB audio headset)

    defaults which match the vendor-specified power-on reset value (as long as they are independent from other properties)

Power-on reset values may be used for defaults as long as they’re independent. If changing one property would require changing another to create a consistent configuration, then those properties should be made required.

In any case where default: is used, the property documentation should explain why the value was selected and any conditions that would make it necessary to provide a different value. (This is mandatory for built-in bindings.)
```

Here are some snippets from the updated binding with follow-up questions.

```
  offset:
[...]
      Initial offset for the driver.
      Default value set to 0 as this is the identity element.
[...]
```

I guess this is supposed to mean that it is a standard initial configuration? I think `DT_INST_PROP_OR` as you suggested in an earlier comment would probably be more appropriate, but I'm not sure, because the definition of "offset" is circular: the offset is the "initial offset", but it's not clear what that means. Offset of what?

Referencing a driver is also a layering issue which I'll get into more detail about lower on:

```
      Can be changed during runtime using sensor_attr_set(dev,SENSOR_CHAN_WEIGHT,SENSOR_ATTR_OFFSET,val)
      and calculated using avia_hx711_tare().
```

Devicetree properties should be used as hardware description and initial configuration, so software-level concerns like specific zephyr driver API calls are not appropriate to put in devicetree property descriptions.

I know we don't have a good place to describe individual zephyr device drivers and I appreciate what you are trying to do here. I know that devicetree bindings can be a tempting place to put this sort of useful information, but it's not an appropriate layer in the software stack for it; please don't do this.

```
  slope:
[...]
    default: [0 0]
    description: |
      Initial slope for the driver.
      Default value set to 1.0 as this is the identity element.
      Can be changed during runtime using sensor_attr_set(dev,SENSOR_CHAN_WEIGHT,SENSOR_ATTR_SLOPE,val)
      and calculated using avia_hx711_calibrate()
```

The same comments apply to `slope` as as applied to `offset`:

- circular definition in the property description is unclear and makes it hard for me to understand if the default justification is appropriate
- layering problems between drivers and DT


```
  calibration-weight:
    type: array
    required: false
    default: [0 0]
    description: |
      Default calibration weight.
      Default value set to 0.0 as this will cause hx711_channel_get(dev,SENSOR_CHAN_WEIGHT,val) to always return 0 indicating that
      calibration weight has not been provided either in DT or during runtime.
      Can be changed during runtime using sensor_attr_set(dev,SENSOR_CHAN_WEIGHT,SENSOR_ATTR_CALIBRATION,val)
```

Same comments here.

```
  sample-fetch-timeout-ms:
    type: int
    required: false
    default: 700
    description: |
      Time until sample-fetch will wait before returning an error.
      Most module implementations have a wake-up time after power up of about 600ms.
      Default set to 700ms to ensure most implementations will work.
      Increase if necessary.
```

This is perfect. This is a textbook case of "delays that would be different only under unusual conditions (such as intervening hardware)". Nice job with this one.

```
  gain:
    type: int
    required: false
    default: 1
    description: |
      Gain of the sensor. 1 = 128x, 2 = 32x, 3 = 64x.
      Default set to 1(128x) as this is the sensors default setting after power-up.
      Can be changed during runtime using sensor_attr_set(dev,SENSOR_CHAN_WEIGHT,SENSOR_ATTR_GAIN,val)
    enum:
      - 1
      - 2
      - 3
```

Almost perfect! This just has a layering issue since it references specific driver APIs. That should be removed, but the rest is an excellent justification under the "defaults which match the vendor-specified power-on reset value" criterion.

```
  rate-hz:
    type: int
    required: false
    default: 0
    description: |
      Sensor sample rate. 0 = 10Hz, 1 = 80Hz
      Default set to 0 (10Hz). Can vary depending on implementation.
      If CONFIG_PM_DEVICE_RUNTIME = y rate will not be higher
      than 2~3Hz due to the long wake-up time needed for HX711.
      Can be changed during runtime using sensor_attr_set(dev,SENSOR_CHAN_WEIGHT,SENSOR_ATTR_SAMPLING_FREQUENCY,val)
    enum:
      - 0
      - 1
```

To me this default seems risky (as in "risk in using default: when the value in the binding may be incorrect for a particular board or hardware configuration" from the requirements).

Are you sure this shouldn't just be made required, "forcing the devicetree maintainer into an explicit and witting choice"?

nobodyguy commented:
@GeorgeGkinis I adore your patience with the constant loop of rebasing and fixing stuff that has diverged.

GeorgeGkinis commented:
@MaureenHelm 
Comments are added to the code.

Tasosgi commented:
I think it must be placed at drivers/ADC, not drivers/sensor

nobodyguy commented:
> I think it must be placed at drivers/ADC, not drivers/sensor

That's a matter of perspective. From my point of view, general ADC driver should be in drivers/ADC.
But I'd put ADC driver customized for weight scale applications to drivers/sensor.
I guess the current state is OK since this concern hasn't been said by any of the core contributors yet.

mbolivar-nordic commented on samples/sensor/hx711/README.rst:
```diff
@@ -0,0 +1,79 @@
+.. _hx711:
+
+HX711: 24-Bit Analog-to-Digital Converter (ADC) for Weigh Scales
```
The commit message is about adding a driver, but this is adding an application. Please update.

GeorgeGkinis replied:
done.

mbolivar-nordic commented on dts/bindings/sensor/avia,hx711.yaml:
```diff
@@ -0,0 +1,66 @@
+    description: |
+      Pin on which sampling rate is chosen.
+
+  label:
+    required: true
```
Please remove this part. @galak  is on a crusade right now to fix this misbehavior throughout the tree. The label property is never required for sensor devices initialized from DT nodes.

galak replied:
As @mbolivar-nordic I'm on a crusade.  So please also remove from any .dts{i} & .overlay files.

GeorgeGkinis replied:
removed.

mbolivar-nordic reviewed (changes requested):
Thanks for your patience. I noticed a problem with the commit log while I was skimming this, and I have one more request for the binding, but otherwise the binding looks great now, thank you!

galak commented on samples/sensor/hx711/boards/nrf52dk_nrf52832.overlay:
```diff
@@ -0,0 +1,20 @@
+	hx711:hx711 {
+			compatible = "avia,hx711";
+			status = "okay";
+			label = "HX711";
```
remove `label` here.

GeorgeGkinis replied:
removed.

galak commented on tests/drivers/build_all/sensor/app.overlay:
```diff
@@ -126,6 +126,16 @@
+		hx711 {
+			compatible = "avia,hx711";
+			status = "okay";
+			label = "HX711";
```
remove `label` here

GeorgeGkinis replied:
removed.

MaureenHelm commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,626 @@
+
+struct hx711_config {
+	int instance;
+	// const int32_t cw[2];
+	// const int32_t sl[2];
```
Remove dead code

GeorgeGkinis replied:
Done,

GeorgeGkinis commented:
Hi all!
Thanks again for your feedback.

I am on vacation now and will resume work on this PR after 18 Aug.

GeorgeGkinis commented:
Hello @MaureenHelm I am back from vacations.

Could you run the last tests and verify?

Thanks

GeorgeGkinis commented on drivers/sensor/hx711/hx711.c:
```diff
@@ -0,0 +1,626 @@
+#include <pm/device.h>
+#include <pm/device_runtime.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
```
Done.

GeorgeGkinis commented on include/drivers/sensor/hx711.h:
```diff
@@ -0,0 +1,63 @@
+#endif
+
+#include <device.h>
+#include <drivers/sensor.h>
```
Done.

GeorgeGkinis commented on samples/sensor/hx711/src/main.c:
```diff
@@ -0,0 +1,94 @@
+#include <pm/device.h>
+#include <pm/pm.h>
+#include <sys/reboot.h>
+#include <sys/util.h>
```
Done.

galak commented on drivers/sensor/hx711/Kconfig:
```diff
@@ -0,0 +1,9 @@
+
+config HX711
+	bool "HX711 Weight Scale Sensor"
+	depends on GPIO
```
add:

```
default y
depends on DT_HAS_AVIA_HX711_ENABLED
```

GeorgeGkinis replied:
Added.

galak commented on dts/bindings/sensor/avia,hx711.yaml:
```diff
@@ -0,0 +1,63 @@
+  sample-fetch-timeout-ms:
+    type: int
+    required: false
+    default: 700
```
Not sure if `default` makes sense here since this value varies board to board.

GeorgeGkinis replied:
maybe increase it to 1000ms?
Seems sensible to add a "safe" default value so that the user does not have to attach a oscilloscope or logic analyser to find the value specific for his board.

If puts the effort of measuring his specific solution he can fine-tune.

If you insist of forcing the user to add this then ok.

Please verify.

GeorgeGkinis committed:
sensor: add driver for HX711

Based on Avia Semiconductor's
patented technology, HX711 is a precision 24-bit
analog-to-digital converter (ADC) designed for weigh scales
and industrial control applications to interface directly with
a bridge sensor.

Signed-off-by: Georgios Gkinis <gkinis.georgios@gmail.com>

GeorgeGkinis commented:
Had some inconsistencies between C-docs and method signature.

GeorgeGkinis committed:
sensor: add example for HX711 driver

Simple example using the HX711 driver.

Signed-off-by: Georgios Gkinis <gkinis.georgios@gmail.com>

GeorgeGkinis commented:
@MaureenHelm Could you please point to the documentation regarding testing documentation builds?
There were failed builds on the example's README.rst which were noticed after you approved the remaining workflows.
 I would like to be able to catch them before pushing.

GeorgeGkinis committed:
Merge branch 'main' into hx711-driver

GeorgeGkinis committed:
Merge branch 'main' into hx711-driver

nobodyguy commented:
For those looking for an HX711 driver, here is a out-of-tree version https://github.com/nobodyguy/HX711_zephyr_driver

p-sousa commented:
Hi @GeorgeGkinis any reason for this work to be abandoned? Would be nice to have this merged. Let me know if I can help re-opening this, or if you mind trying to have it merged myself.  Thank you!

nobodyguy commented:
> Hi @GeorgeGkinis any reason for this work to be abandoned? Would be nice to have this merged. Let me know if I can help re-opening this, or if you mind trying to have it merged myself. Thank you!

What's wrong with the out-of-tree version?

GeorgeGkinis commented:
@p-sousa ,

I have dedicated an enormous amount of time trying to get this accepted and merged.

I still have interest contributing to the community, but the (hobby) project that required it has no need for it anymore.

Every time I stumbled across corner cases on merge requirements as can be seen in the gargantuan discussion.

If the community still requires it I could start again, but would like some support.

I am currently on vacation. 

Setting up the HW will require some additional time.

In the meantime I second @nobodyguy 's last comment about using his module.

jurenat commented:
Hi @GeorgeGkinis , 
I would like to see this merged to the upstream. I have already rebased your code on the current main, which is compilable with small fixes.
I am considering using HX712 for my next project (which is almost the same ADC) so I would like to add a more generic version to support it and maybe more variants (like hx717, etc.).
Are you still open to working on this again?

GeorgeGkinis commented:
Hi @jurenat ,

How do you suggest we work this out?
Is there anything specific I could focus on?

Could you share a repo/branch?
All those questions.. 😀

jurenat commented:
I was not expecting such a fast response xD
Well, it depends if you want to get involved but first I would like to clean the code a little and open discussion about the necessity of the `SENSOR_ATTR_SLOPE` symbol being added to API. We can discuss this on Discord.
I can push the rebased and slightly modified version in the evening or tomorrow.

nobodyguy commented:
I'd suggest take a look into our driver module while upstreaming it, there are a few not so obvious fixes like this one - https://github.com/nobodyguy/HX711_zephyr_driver/commit/d3fc3555d15eb96e838284bfdcd89d6497753004

jurenat commented:
@nobodyguy Cool, thanks. I will check that.
@GeorgeGkinis I have pushed two branches to my fork. One is your branch rebased to the main with some small changes, and the second one is my initial work on the more generic version. As I said I do not like the idea of adding the SLOPE attribute to API so I removed it and instead used the CALIBRATION attribute (I do not see a reason to have calibration weight stored in the driver data).
